ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/JmsXAResourceRecoveryHelper.java

```diff
@@ -82,31 +71,31 @@ public boolean initialise(String properties) {
      */
     @Override
     public XAResource[] getXAResources() {
-        if (connect()) {
-            if (jtaLogger.logger.isTraceEnabled()) {
-                jtaLogger.logger.trace("Returning XA resource: " + this);
+        if (!connectionManager.isConnected()) {
+            try {
+                connectionManager.connect();
+            } catch (XAException ignored) {
+                return new XAResource[0];
             }
-
-            return new XAResource[] { this };
         }
 
-        return new XAResource[0];
+        return new XAResource[] { this };
     }
 
     /**
      * Delegates XAResource#recover call to the connected JMS resource. If provided argument is XAResource.TMENDRSCAN, then JMS
      * connection will be closed at the end of the call.
      *
-     * @param i
+     * @param flag
      * @throws XAException
      */
     @Override
-    public Xid[] recover(int i) throws XAException {
+    public Xid[] recover(int flag) throws XAException {
         try {
-            return delegate.recover(i);
+            return connectionManager.connectAndApply(delegate -> delegate.recover(flag));
         } finally {
-            if (i == XAResource.TMENDRSCAN) {
-                disconnect();
+            if (flag == XAResource.TMENDRSCAN) {
+                connectionManager.disconnect();
```

Comment:
I wonder if there is any real benefit to closing the connection at the ENDRSCAN anyway? Reason being that an open idle connection should not take much resource compared to open/closing for the cycle. But I am OK with the change if it addresses the problem. 

I think https://issues.jboss.org/browse/JBTM-2903 should be updated with a stack trace showing an exception coming out of commit though.